Max Accounts and other improvements

BadWolf0081 opened the pull request:
Added a field in config.yml for max accounts per domain, set to 999999 if you dont' want to use it.

Also added chart at the beginning of generation showing accounts per domain.

Also updated rotation logic to handle unknown errors and not rotate on max domain situation only.

BadWolf0081 committed:
Max Accounts and other improvements

Added a field in config.yml for max accounts per domain, set to 999999 if you dont' want to use it.

Also added chart at the beginning of generation showing accounts per domain.

Also updated rotation logic to handle unknown errors and not rotate on max domain situation only.

SpielerNogard commented on ptc/account_generation/config.py:
```diff
@@ -66,6 +66,7 @@ class Config(BaseModel):
     additional_sleep_time: int = 0
     webdriver_max_wait: int = 30
     network_blocks: Optional[List[str]] = None
+    max_accounts_per_domain: int = 100  # Default max accounts per domain
```
not a big fan of this default, default should be no limit (-1 or something)

BadWolf0081 replied:
I didn't realize it had a fallback default.  I can set it to 999999 like I suggested in the PR.  Or do you think people will make more than 1 million per domain?

SpielerNogard replied:
make the default to -1 
and before this part of the code

```python
max_accounts = config.max_accounts_per_domain
if not check_domain_account_limit(domain, max_accounts, accounts_collection):
        logger.error(f"Domain '{domain}' has reached the maximum allowed accounts ({max_accounts}). No account will be created.")
        return False
```

add this:

```python
max_accounts = config.max_accounts_per_domain
if max_accounts:
    if not check_domain_account_limit(domain, max_accounts, accounts_collection):
        logger.error(f"Domain '{domain}' has reached the maximum allowed accounts ({max_accounts}). No account will be created.")
        return False

genomeprjct replied:
instead of erroring when the account reaches limit, could it just try a different domain?

likewise, similar to the `show_total_accounts` config option, would it make sense to print the summary by account on start up if configured?

BadWolf0081 replied:
This should be handled with my latest push.

BadWolf0081 committed:
update max account default

changed from 100 to 999999 as suggested in PR notes

matt-yml commented:
can the "accounts per domain" chart be moved to a debug level of logging?

SpielerNogard commented:
> can the "accounts per domain" chart be moved to a debug level of logging?

yes he just need to replace print with logger.debug

SpielerNogard commented:
> > can the "accounts per domain" chart be moved to a debug level of logging?
> 
> yes he just need to replace print with logger.debug

nvm the logger is internaly just a print with different colors, so the complete logging need to be changed for that, no big thing theoreticaly 🙃

matt-yml commented:
> nvm the logger is internaly just a print with different colors, so the complete logging need to be changed for that, no big thing theoreticaly 🙃

i ran into that issue when adding other log stuff to mail_server (that i've been testing over the last week)
instead of augmenting project-wide logging i tried to KISS and just cleaned up my log messages. `:)`

genomeprjct commented on ptc/account_generation/index.py:
```diff
@@ -423,6 +470,18 @@ def generate(
+    display_domain_account_chart(accounts_collection)
+
+    # Check domain account limit
+    max_accounts = config.max_accounts_per_domain
```
If I'm reading this block right, you're going to print this chart on every generation, which is likely going to be too noisy.

matt-yml replied:
the chart is based on the local mongodb. this is good because it avoids the complexity of checking against the main drago or rdm table. 
however, some users run AIO and have a db per script instance. in my setup i use a single backend for all scripts and hosts. 

if the goal is to measure the global count for each domain we get a more useful result at the mysql/mariadb main.

this chart can be run independently of account_generation, is provider agnostic, and be added to a workflow by user choice. 
```sql
-- Set the provider to filter
-- e.g. ptc, nk, g, or NULL for all
SET @provider_filter = 'ptc' COLLATE utf8mb4_unicode_ci;

SELECT 
    -- Extract the root domain from the email address
    CASE 
        WHEN email IS NULL THEN 'NULL' 
        ELSE SUBSTRING_INDEX(SUBSTRING_INDEX(email, '@', -1), '.', -2) 
    END AS root_domain,
    
    -- Count the occurrences of each domain
    COUNT(*) AS domain_count

FROM account

-- Apply the provider filter if specified
WHERE (@provider_filter IS NULL OR provider COLLATE utf8mb4_unicode_ci = @provider_filter)

-- Group by the extracted domain to aggregate counts
GROUP BY root_domain

-- Optionally, order the results by count in descending order
ORDER BY domain_count DESC;
```

genomeprjct replied:
You're assume the user of this script is loading the accounts directly into the target DB rather than generating and importing them separately.  Either way, I think it makes the point that the report of accounts per domain should be independent of the processing

i actually do it via a script I run against the local account files that got generated, but to the point of this PR equally good to just have it point at mongo

genomeprjct replied:
Also, I just tested out this report against my local data.  It doesn't generate what I consider to be the right output since I ran with subdomains for a little bit.  It actually prints one row per subdomain, instead of grouping based on the top level domains.

matt-yml replied:
agreed. others also may import to different controller instances, databases, or tables. 
that's why i submitted that sql so people can run it appropriately for their use case.

refocusing on this PR: 
yes. keeping it simple and scoped to the single mongodb instance works well.

i like how our shared interest to prevent logs from getting noisy is satisfied by your idea to only print a domain report chart at account_generation start.
i was avoiding the noise by offering an external sql script replacement, but your approach is better for the PR.

BadWolf0081 committed:
Fix to make chart optional

set show_chart to false in config.yml to disable chart

BadWolf0081 committed:
Fixes

Changed so if max accounts is omited from config it now is unlimited (-1)

Changed so if no domains are under their limit the script will close.
